File: parasmittal/Day 4/Count Binary Substrings.cpp
class Solution {
public:
    int countBinarySubstrings(string s) {
        int n=s.length();
        if(n<=1)
            return 0;
       int countzero=0,countone=0;
        int ans=0;
        if(s[0]=='1')
            countone++;
        else
            countzero++;
            
        for(int i=1;i<n;i++){
            if(s[i]!=s[i-1]){
                if(s[i]=='1'){
                    countone=1;
                }
                else{
                    countzero=1;
                }
                ans++;
            }
            else{
                 if(s[i]=='1')
            countone++;
                 
        else
            countzero++;
                if(countzero==countone ||(s[i]=='1' && countzero>countone) || (s[i]=='0' && countone>countzero)){
                ans++;
                }
            }
        }
        return ans;
    }
};

File: parnamondal/Week - 1/Day - 4/COUNT BINARY SUBSTRINGS.py
# Time Complexity - O(N)
# Space Complexity - O(N)

import itertools
class Solution:
    def countBinarySubstrings(self, s: str) -> int:
        n=len(s)
        if n<=1: 
            return 0
        if n==2:
            if s=="01" or s=="10":
                return 1
            else:
                return 0
        c=0
        a=[ len(list(g)) for c,g in itertools.groupby(s)]
        for i in range (len(a)-1):
            c+=min(a[i],a[i+1])
        return c
            